mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc

```diff
@@ -0,0 +1,228 @@
+  const float yaw = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[0];
+  const float pitch = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[1];
+  const float roll = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[2];
+
+  for (const Anchor &anchor : translation_data) {
+    int id = anchor.sticker_id;
+
+    TimedModelMatrixProto* model_matrix =
+        model_matrix_list->add_model_matrix();
+    model_matrix->set_id(id);
+
+    float rotation = getUserRotation(user_rotation_data, id);
+    float scaling = getUserScaling(user_scaling_data, id);
+
+    const Eigen::Vector3f translation_vector = generateTranslationVector(anchor, scaling);
+    const Eigen::Matrix3f rotation_submatrix = generateRotationSubmatrix(yaw, pitch, roll, rotation);
+
+    Matrix4fRM mvp_matrix = generateEigenModelMatrix(translation_vector, rotation_submatrix);
+
+    // The generated model matrix must be mapped to TimedModelMatrixProto (col-wise)
+    for (int i = 0; i < mvp_matrix.rows(); ++i) {
+      for (int j = 0; j < mvp_matrix.cols(); ++j) {
+        model_matrix->add_matrix_entries(mvp_matrix(j, i));
+      }
+    }
+  }
+
+  cc->Outputs()
+      .Tag(kModelMatricesTag)
+      .Add(model_matrices.release(), cc->InputTimestamp());
+
+  return ::mediapipe::OkStatus();
+}
+
+// Using the tracked translation data, generate a vector for MVP translation
+Eigen::Vector3f MatricesManagerCalculator::generateTranslationVector(Anchor tracked_anchor, float user_scaling_increment) {
+  // Convert from normalized [0.0-1.0] to openGL on-screen coordinates
+  float z = (initial_z_ + user_scaling_increment) * tracked_anchor.z;
+
+  // TODO: Investigate possible differences in warping of tracking speed across screen
+  float y_minimum = z * (tan(vertical_fov_radians_ /2));
+  // Minimum y value appearing on screen at z distance
+  float x_minimum = y_minimum * (1.0/aspect_ratio_);
+  // Minimum x value appearing on screen at z distance
+
+  float x = (tracked_anchor.x * (-x_minimum - x_minimum)) + x_minimum;
+  float y = (tracked_anchor.y * (-y_minimum - y_minimum)) + y_minimum;
+
+  Eigen::Vector3f t_vector(x,y,z);
+  return t_vector;
+}
+
+// Generate the submatrix defining rotation using IMU data and user rotations
+Eigen::Matrix3f MatricesManagerCalculator::generateRotationSubmatrix(float yaw, float pitch, float roll, float user_rotation_radians) {
+  Eigen::Matrix3f r_submatrix;
+  r_submatrix =
+      Eigen::AngleAxisf(yaw, Eigen::Vector3f::UnitY()) *
+      Eigen::AngleAxisf(-roll, Eigen::Vector3f::UnitZ()) *
+      Eigen::AngleAxisf(pitch - M_PI/2, Eigen::Vector3f::UnitX());
+  return r_submatrix;
+}
+
+// Generates a model matrix via Eigen with appropriate transformations
+Matrix4fRM MatricesManagerCalculator::generateEigenModelMatrix(
+    Eigen::Vector3f translation_vector, Eigen::Matrix3f rotation_submatrix) {
+  // Define basic empty model matrix
+  Matrix4fRM mvp_matrix;
+
+  // Set the translation vector
+  mvp_matrix.bottomLeftCorner<1, 3>() = translation_vector;
```

Comment:
interesting, I thought the translation is at the last column?

Reply:
The method of model matrices generation performed in GlCalc did all the matrix generation and then transposed to an output matrix. However, for simplicity's sake, I will modify this to remove the need for transpose (instead of making it the bottom row, it will be the last column).